File: tipos-suma/sum.ts


type Position = {
    x: number,
    y: number,
};

interface Circle {
    type: "circle",
    position: Position,
    radius: number,
}

type Rect = {
    type: "rect",
    position: Position,
    width: number,
    height: number,
}

type Shape = Circle | Rect;

function middlePoint(a: Shape, b: Shape): Position {
    const x = (a.position.x + b.position.x) / 2;
    const y = (a.position.y + b.position.y) / 2;
    return { x, y};
}

function area(a: Shape, b: Shape): number {
    if(a.type === "circle"){
        return Math.PI*Math.pow(a.radius, 2);
    } else {
        return a.height*a.width;
    }
}

function main(){
    let shapes: Shape[] = [];
    shapes.push({
        type: "circle",
        position: {
            x: 50,
            y: 60
        },
        radius: 7
    });
    shapes.push({
        type: "rect",
        position: {
            x: 50,
            y: 60,
        },
        width: 40,
        height: 30,
    });
    const point = middlePoint(shapes[0], shapes[1]);


    // NULL
    let shape = shapes.find((t: Shape) => t.position.x === 40);
    if(shape){
        area(shape, shapes[0]);
    }

}

main();

File: tipos-suma/sum.rs
struct Position {
    x: f64,
    y: f64,
}

enum Shape {
    Circle {
        position: Position,
        radius: f64,
    },
    Rect {
        position: Position,
        width: f64,
        height: f64,
    }
}

fn area(shape: &Shape) -> f64 {
    match shape {
        Shape::Circle { radius, .. } => std::f64::consts::PI * (radius.powf(2.0)),
        Shape::Rect { width, height, ..} => width * height
    }
}

fn area2(shape: &Shape) -> f64 {
    if let Shape::Circle { radius, .. } = shape {
        return std::f64::consts::PI * (radius.powf(2.0));
    }
    if let Shape::Rect { width, height, ..} = shape {
        return width * height;
    }
    unreachable!();
}

fn main(){
    let mut shapes = Vec::new();
    shapes.push(Shape::Circle {
        position: Position {
            x: 5.0,
            y: 6.0,
        },
        radius: 4.0
    });
    shapes.push(Shape::Rect {
        position: Position {
            x: 5.0,
            y: 6.0,
        },
        width: 70.0,
        height: 80.0
    });

    let shape = shapes.get(0);
    if let Some(shape) = shape {
        area(&shape);
    }
    
}